packages/@sanity/schema/src/sanity/extractSchema.ts

```diff
@@ -0,0 +1,669 @@
+      })
+      return
+    }
+
+    if (isReferenceType(type)) {
+      baseTypes.push({
+        type: 'type',
+        name: type.name,
+        value: createReferenceFieldDefintion(type),
+      })
+      return
+    }
+
+    if (isCrossDatasetReferenceType(type)) {
+      baseTypes.push({
+        type: 'type',
+        name: type.name,
+        value: createCrossDatasetReferenceFieldDefintion(type),
+      })
+      return
+    }
+
+    if (typeMap[type.type]) {
+      baseTypes.push({
+        type: 'type',
+        name: type.name,
+        value: typeMap[type.type],
+      })
+      return
+    }
+
+    baseTypes.push({
+      type: 'type',
+      name: type.name,
+      value: {
+        type: 'reference',
+        to: type.type,
+      } satisfies ReferenceField,
+    })
+  })
+
+  return baseTypes
+}
+
+function createKeyField(): ObjectKeyValueField<OptionalField<StringField>> {
+  return {
+    type: 'objectKeyValue',
+    key: '_key',
+    value: optionalField({
+      type: 'string',
+    }),
+  }
+}
+
+function optionalField<T extends Field = Field>(field: T): OptionalField<T> {
+  return {
+    type: 'optional',
+    value: field,
+  }
+}
+function parseField(field: Exclude<FieldDefinition, SchemaField>): Field {
+  if (isObjectType(field)) {
+    const fields: ObjectKeyValueField[] = []
+    field.fields.forEach((f) => {
+      fields.push({
+        type: 'objectKeyValue',
+        key: f.name,
+        value: optionalField(parseField(f)),
+      })
+    })
+    fields.push({
+      type: 'objectKeyValue',
+      key: '_type',
+      value: {
+        type: 'string',
+        value: field.name,
+      },
+    } as ObjectKeyValueField<StringField>)
+
+    return {
+      type: field.type,
+      fields,
+    }
+  }
+
+  if (isArrayType(field)) {
+    return createArray(field)
+  }
+
+  if (isBlockType(field)) {
+    return createBlock(field)
+  }
+  if (isImageType(field)) {
+    return createImage(field)
+  }
+  if (isFileType(field)) {
+    return createFile(field)
+  }
+
+  if (isReferenceType(field)) {
+    return createReferenceFieldDefintion(field)
+  }
+
+  if (isCrossDatasetReferenceType(field)) {
+    return createCrossDatasetReferenceFieldDefintion(field)
+  }
+
+  if (typeMap[field.type]) {
+    return typeMap[field.type]
+  }
+
+  return {
+    type: 'reference',
+    to: field.type,
+  }
+}
+
+function isDocumentType(n: SchemaTypeDefinition): n is DocumentDefinition {
+  return n.type === 'document'
+}
+function isFieldDefinition(n: unknown): n is FieldDefinition {
+  return (
+    n !== null &&
+    typeof n === 'object' &&
+    'type' in n &&
+    (('fieldset' in n && typeof n.fieldset === 'string') ||
+      !('fieldset' in n) ||
+      typeof n.fieldset === 'undefined') &&
+    (('group' in n && (typeof n.group === 'string' || Array.isArray(n.group))) ||
+      !('group' in n) ||
+      typeof n.group === 'undefined')
+  )
+}
+
+function isObjectType(n: {type: string}): n is ObjectDefinition {
```

Comment:
a complicating factor here is that we support aliasing types with the schema today, so you can do:

```ts
const schemaTypes = [
  {type: 'object', name: 'myObject', fields: [{type: 'string', name: 'title'}]},
  {type: 'myObject', name: 'mySubObject'},
]
```
Here, myObject will not pass as an object (I think?), and you will need to use the schema to walk upwards in the type chain, to correctly identify it as an object.

Note: We don't allow extending fields, so `mySubObject` is structurally similar to `mySubObject`, but it's values will have a different `_type`.

I guess the end result here will be that the value type of `mySubObject` will be unknown, which might be a perfectly fine trade-off, just want you to be aware.

Reply:
I guess here it would fail to assert the type and fallback to:
```
return {
    type: 'reference',
    to: field.type,
  }```
  
  Which would be correct? 🤔